engine/src/main/java/org/pentaho/di/shared/SharedObjectsManager.java

```diff
@@ -0,0 +1,223 @@
+    switch ( SharedObjectType.valueOf( type.toUpperCase() ) ) {
+      case CONNECTION:
+        populateDbMetas();
+        break;
+      case SLAVESERVER:
+        populateSlaveServers();
+        break;
+      case PARTITIONSCHEMA:
+        // populate partitionSchemas
+        break;
+      case CLUSTERSCHEMA:
+        //populate clusterSchemas
+        break;
+      default:
+        log.error( " Invalid Shared Object type " + type );
+        throw new KettleXMLException( "Invalid shared object type " + type );
+    }
+  }
+
+  /**
+   * Populate the SharedObjectsMap. The map is of the form {type, Map{name, SharedObjectInterface}}.
+   * @throws KettleException
+   */
+  private void populateDbMetas() throws KettleException {
+    if ( !dbconn_initialized ) {
+      synchronized ( this ) {
+        if ( !dbconn_initialized ) {
+          Map<String, Node> nodeMap = sharedObjectsIO.getSharedObjects( DB_TYPE );
+          Map<String, SharedObjectInterface> dbSharedObjectMap = new HashMap<>();
+          for ( String name : nodeMap.keySet() ) {
+            SharedObjectInterface dbSharedObject = new DatabaseMeta( nodeMap.get( name ) );
+            if ( !dbSharedObjectMap.containsKey( name ) ) {
+              dbSharedObjectMap.put( name, dbSharedObject );
+            }
+          }
+          sharedObjectsMap.put( DB_TYPE, dbSharedObjectMap );
+        }
+      }
+      dbconn_initialized = true;
+    }
+  }
+
+  private void populateSlaveServers() throws KettleException {
+    if ( !slaveServer_initialized ) {
+      synchronized ( this ) {
+        if ( !slaveServer_initialized ) {
+          Map<String, Node> nodeMap = sharedObjectsIO.getSharedObjects( SLAVESERVER_TYPE );
+          Map<String, SharedObjectInterface> slaveServerMap = new HashMap<>();
+          for ( String name : nodeMap.keySet() ) {
+            SharedObjectInterface slaveServerSharedObject = new SlaveServer( nodeMap.get( name ) );
+            if ( !slaveServerMap.containsKey( name ) ) {
+              slaveServerMap.put( name, slaveServerSharedObject );
+            }
+          }
+          sharedObjectsMap.put( SLAVESERVER_TYPE, slaveServerMap );
+        }
+      }
+      slaveServer_initialized = true;
+    }
+  }
+
+  /**
+   * Save the SharedObjectInterface for a type using the persistence mechanism defined by SharedObjectIO and
+   * to the local map
+   * @param type The type of sharedObject
+   * @param sharedObjectInterface
+   * @throws KettleException
+   */
+  @Override
+  public synchronized void add( String type, SharedObjectInterface sharedObjectInterface )  throws KettleException {
+    populateSharedObjectMap( type );
+    String name = sharedObjectInterface.getName();
+    Node node = toNode( type, sharedObjectInterface );
+    sharedObjectsIO.saveSharedObject( type, name, node );
+
+    // Add it to the map
+    Map<String, SharedObjectInterface> localSharedObjectsMap = sharedObjectsMap.get( type );
```

Comment:
doesn't this need to computeIfAbsent? 

the calls to computeIfAbsent should maybe be refactored out to a helper method.